sql/src/influxql/provider.rs

```diff
@@ -0,0 +1,466 @@
+    /// and nullable
+    fn new(schema: &Schema) -> Result<Self> {
+        let cols = schema.columns();
+        let timestamp_key_idx = schema.timestamp_index();
+        let tsid_idx_opt = schema.index_of_tsid();
+        let arrow_fields = &schema.to_arrow_schema_ref().fields;
+
+        let influx_columns = arrow_fields
+            .iter()
+            .zip(cols.iter().enumerate())
+            .filter_map(|(arrow_col, (col_idx, col))| {
+                if matches!(tsid_idx_opt, Some(tsid_idx) if col_idx == tsid_idx) {
+                    None
+                } else {
+                    let influx_type_res =
+                        map_column_type_to_influx_column_type(col, col_idx == timestamp_key_idx);
+                    Some(influx_type_res.map(|influx_type| (influx_type, arrow_col.clone())))
+                }
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        Ok(Self {
+            columns: influx_columns,
+        })
+    }
+}
+
+impl InfluxSchema for InfluxSchemaImpl {
+    fn columns(&self) -> Vec<(InfluxColumnType, &ArrowField)> {
+        self.columns
+            .iter()
+            .map(|(influx_type, field)| (*influx_type, field))
+            .collect()
+    }
+
+    fn tags(&self) -> Vec<&ArrowField> {
+        self.columns
+            .iter()
+            .filter_map(|(influx_type, field)| {
+                if matches!(influx_type, InfluxColumnType::Tag) {
+                    Some(field)
+                } else {
+                    None
+                }
+            })
+            .collect()
+    }
+
+    fn fields(&self) -> Vec<&ArrowField> {
+        self.columns
+            .iter()
+            .filter_map(|(influx_type, field)| {
+                if matches!(influx_type, InfluxColumnType::Field(..)) {
+                    Some(field)
+                } else {
+                    None
+                }
+            })
+            .collect()
+    }
+
+    fn time(&self) -> &ArrowField {
+        // Time column must exist, has checked it when building.
+        let time_column = self
+            .columns
+            .iter()
+            .find(|(influx_type, _)| matches!(influx_type, InfluxColumnType::Timestamp))
+            .unwrap();
+
+        &time_column.1
+    }
+
+    fn column(&self, idx: usize) -> (InfluxColumnType, &ArrowField) {
+        let (influx_type, field) = &self.columns[idx];
+
+        (*influx_type, field)
+    }
+
+    fn find_index_of(&self, name: &str) -> Option<usize> {
+        self.columns
+            .iter()
+            .enumerate()
+            .find(|(_, (_, field))| field.name() == name)
+            .map(|(index, _)| index)
+    }
+}
+
+fn map_column_type_to_influx_column_type(
+    column: &ColumnSchema,
+    is_timestamp_key: bool,
+) -> Result<InfluxColumnType> {
+    // Time column
+    if is_timestamp_key {
+        if column.name == "time" && !column.is_nullable {
```

Comment:
I prefer to return early to avoid too many `else if` clause.

Reply:
Why this column must be `time`? in our codebase default ts column name is `timestamp`, not `time`.

I think the only requirement is there must exists a timestamp column in schema, no name constraint.